src/base.ml

```diff
@@ -50,101 +48,132 @@ let render_texture (r : Sdl.renderer) (texture : Sdl.texture) (s : Screen.t) (bi
+      (match Sdl.create_texture r Sdl.Pixel.format_rgb888 ~w:width ~h:height
+              Sdl.Texture.access_streaming with
+       | Error (`Msg e) ->
+           Sdl.log "Texture error: %s" e;
+           exit 1
+       | Ok texture ->
+           let bitmap =
+             Bigarray.Array1.create Bigarray.int32 Bigarray.c_layout (width * height)
+           in
+           let initial_buffer =
+             match boot with
+             | None -> Framebuffer.init (width, height) (fun _ _ -> 0)
+             | Some bfunc -> bfunc s
+           in
+           let initial_input = {
+             keys = KeyCodeSet.empty;
+             events = [];
+             mouse = Mouse.create scale;
+           } in
+           let rec loop t prev_buffer input last_t =
+             let now = Sdl.get_ticks () in
+             let diff = Int32.sub (Int32.of_int (1000 / 60)) (Int32.sub now last_t) in
+             if Int32.compare diff Int32.zero > 0 then Sdl.delay diff;
+             let exit, new_keys, new_mouse, unified_events =
+               poll_all_events input.keys input.mouse []
+             in
+             let current_input = {
+               keys = new_keys;
+               events = unified_events;
+               mouse = new_mouse;
+             } in
+             if exit then ()
+             else begin
+               let updated_buffer = tick t s prev_buffer current_input in
+               if (updated_buffer != prev_buffer)
+                  || (Framebuffer.is_dirty updated_buffer)
+                  || (Screen.is_dirty s)
+               then (
+                 framebuffer_to_bigarray s updated_buffer bitmap;
+                 (match render_texture r texture s bitmap with
+                  | Error (`Msg e) -> Sdl.log "Render error: %s" e
+                  | Ok () -> ());
+                 Framebuffer.clear_dirty updated_buffer;
+                 Screen.clear_dirty s
+               );
+               (match render_texture r texture s bitmap with
+                | Error (`Msg e) -> Sdl.log "Render error: %s" e
+                | Ok () -> ());
+               loop (t + 1) updated_buffer current_input now
+             end
+           in
+           loop 0 initial_buffer initial_input Int32.zero;
+           Sdl.destroy_texture texture;
+           Sdl.destroy_renderer r;
+           Sdl.destroy_window w;
+           Sdl.quit ())
+
+let run_functional title tick_f s =
+  let wrap_tick t screen prev_framebuffer input =
+    let primitives = tick_f t screen input in
+    if primitives = [] then prev_framebuffer
     else
       let width, height = Screen.dimensions screen in
-      let new_framebuffer = Framebuffer.init (width, height) (fun _x _y -> 0) in
+      let new_framebuffer = Framebuffer.init (width, height) (fun _ _ -> 0) in
       Framebuffer.render new_framebuffer primitives;
       new_framebuffer
   in
   run title None wrap_tick s
-      
+
+(* --- Utility functions for input handling --- *)
+
+let is_key_pressed input key =
```

Comment:
Can you add a few tests for these functions? They clearly work today, but the tests are there to make sure they don't get broken in the future.